File: 백준 문제 풀이 모음/class 3/17626.py
import sys

N = int(sys.stdin.readline())
count = 0

DP = [0] * (N+1)

k = 1
while k ** 2 <= N :
    DP[k ** 2] = 1
    k += 1

for i in range(1, N+1) :
    if DP[i] != 0 :
        continue
    j = 1
    while j* j <= i :
        if DP[i] == 0 :
            DP[i] = DP[j*j] + DP[i - j *j]
        else :
            DP[i] = min(DP[i], DP[j*j] + DP[i - j *j])
        j+=1
print(DP[N])

File: 백준 문제 풀이 모음/kotlin/class 3/17626.kt


fun main() = with(System.`in`.bufferedReader()){
    val  n = readLine().toInt()

    val list = MutableList<Int>(n+1) { 0 }
    var k = 1
    while (k*k <= n) {
        list[k*k] = 1
        k++
    }

    for (i in 1 .. n) {
        if (list[i] != 0) continue

        var j = 1
        while (j*j <= i) {
            if (list[i] == 0) {
                list[i] = list[j*j] + list[i-j*j]
            } else {
                list[i] = Math.min(list[i], list[j*j] + list[i-j*j])
            }

            j++
        }

    }

    println(list)
    println(list[n])




}